### PR TITLE
[Android] Implement OTA Provider

### LINKED ISSUE #29147: [Android] For android OTA support, need to implement BdxOTASender in controller JNI that inherits  bdx::Responder
### Feature description

After https://github.com/project-chip/connectedhomeip/pull/28316, we need further implement BdxOTASender, which would at minimum provide 

class BdxOTASender : public bdx::Responder {
public:
    BdxOTASender() {};
    CHIP_ERROR PrepareForTransfer(FabricIndex fabricIndex, NodeId nodeId)
    CHIP_ERROR Init(System::Layer * systemLayer, Messaging::ExchangeManager * exchangeMgr)
    CHIP_ERROR Shutdown()
    void ResetState()
private:
    static void HandleBdxInitReceivedTimeoutExpired(chip::System::Layer * systemLayer, void * state)
    CHIP_ERROR OnMessageToSend(TransferSession::OutputEvent & event)
    bdx::StatusCode GetBdxStatusCodeFromChipError(CHIP_ERROR err)
    CHIP_ERROR OnTransferSessionBegin(TransferSession::OutputEvent & event)
    CHIP_ERROR OnTransferSessionEnd(TransferSession::OutputEvent & event)
    CHIP_ERROR OnBlockQuery(TransferSession::OutputEvent & event)
    CHIP_ERROR ConfigureState(chip::FabricIndex fabricIndex, chip::NodeId nodeId)
};

### Platform

android

### Platform Version(s)

_No response_

### Anything else?

_No response_

### LINKED ISSUE #29267: [Android]Implement OTAProviderDelegate in JNI and hook with Java Layer
We need to implement OTAProviderDelegate with below functions and hooks with Java controller
void HandleQueryImage(
        chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
        const chip::app::Clusters::OtaSoftwareUpdateProvider::Commands::QueryImage::DecodableType & commandData) override;
    void HandleApplyUpdateRequest(
        chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
        const chip::app::Clusters::OtaSoftwareUpdateProvider::Commands::ApplyUpdateRequest::DecodableType & commandData) override;
    void HandleNotifyUpdateApplied(
        chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
        const chip::app::Clusters::OtaSoftwareUpdateProvider::Commands::NotifyUpdateApplied::DecodableType & commandData) override;

### LINKED ISSUE #29298: [Android]Add support for AnnouceOTAProvider Command via OTARequester cluster
Two options:
1. Manually Implement AnnouceOTAProvider command using Generic IM Command API.
2. Codegen all commands using jinja codegen? Size consideration? Maybe it worths pursuing this way, the benefits lies at the later certification test for android matter controller

### PR DESCRIPTION
**Pull request by joonhaengHeo**
Fix #29147 
Fix #29267 
Fix #29298 

1. Add OTA Provider delegate for Android Platform
2. Add OTA Provider UI in Android ChipTool
3. Implement BdxOTASender for Android Platform

### CODE REVIEW

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (yunhanw-google) [inline src/controller/java/CHIPDeviceController-JNI.cpp:561]: where do we release this otaProviderDelegateRef?

Comment (yunhanw-google) [reply src/controller/java/CHIPDeviceController-JNI.cpp:561]: why do you think we need global ref for this delegate? In invoke and write, we also pass callback via jobject, and they are async operations, it seems the jobject passed from java don't need to be global?

File: src/controller/java/OTAProviderDelegateBridge.h
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.h:64]: I suggest to have JniGlobalReference mOtaProviderDelegate, which is recommended pattern to handle global ref for class member

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (yunhanw-google) [reply src/controller/java/CHIPDeviceController-JNI.cpp:561]: I suggest to have JniGlobalReference mOtaProviderDelegate in OTAProviderDelegate, which is recommended pattern to handle global ref for class member.
see example from https://github.com/project-chip/connectedhomeip/pull/29710/files

**Review by yunhanw-google — CHANGES_REQUESTED** [review]

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (yunhanw-google) [inline src/controller/java/CHIPDeviceController-JNI.cpp:587]: no need for this SuccessOrExit.

**Review by yunhanw-google — COMMENTED** [review]
Please apply this to unblock java CI, thanks
```
--- a/src/controller/java/BUILD.gn
+++ b/src/controller/java/BUILD.gn
@@@ -97,7 -97,16 +97,10 @@@ shared_library("jni")
    }
    if (chip_build_controller_dynamic_server) {
+     defines += [
+       "CHIP_DEVICE_CONFIG_DYNAMIC_SERVER" ,
+       "CHIP_CONFIG_SKIP_APP_SPECIFIC_GENERATED_HEADER_INCLUDES=1",
+     ]
      deps += [ "${chip_root}/src/controller:nodatamodel" ]
    }
```

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt
Comment (yunhanw-google) [inline examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt:44]: should we consider to add UI to set this vendor ID so companies can quicky play around the vendorId for testing purpose?

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt
Comment (yunhanw-google) [inline examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt:118]: For readability, I generally prefer to reduce the brace, and return early if possible and avoid !! if possible
if (contentUri.path!!.startsWith("/storage")
  return contentUri.path!!
val id = ....

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt
Comment (yunhanw-google) [inline examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt:280]: Please try to avoid !! across this file.

File: src/controller/java/AndroidDeviceControllerWrapper.cpp
Comment (yunhanw-google) [inline src/controller/java/AndroidDeviceControllerWrapper.cpp:530]: then otaProviderBridge = nullptr.

File: src/controller/java/BdxOTASender.cpp
Comment (yunhanw-google) [inline src/controller/java/BdxOTASender.cpp:195]: UtfString class is problematic, when calling from native layer, UtfString is casting local to global, but cannot release local, which would cause local reference leaks.
The recommended pattern is as below. I have converted this for android callback cpp file, need to spread this pattern to all jni file, at this moment, UTFSting and ByteArray..in JniWrapper should not be used now..
Pattern:
JniLocalReferenceManager manager(env); ==> manage the lifetime for all local refs in current function.
....
jstring fileDesignator = env->NewStringUTF(xx) :

File: src/controller/java/BdxOTASender.cpp
Comment (yunhanw-google) [inline src/controller/java/BdxOTASender.cpp:326]: localRef leaks happens here, please apply the below pattern, 

Pattern:
JniLocalReferenceManager manager(env); ==> manage the lifetime for all local refs in current function.
....
jstring fileDesignator = env->NewStringUTF(xx) 

example code is here, https://github.com/project-chip/connectedhomeip/blob/master/src/controller/java/AndroidCallbacks.cpp#L93C5-L93C43

File: src/controller/java/BdxOTASender.cpp
Comment (yunhanw-google) [inline src/controller/java/BdxOTASender.cpp:328]: please stop using this way to manage the lifetime for localRef, which is leaking.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:182]: boxedLocation is leaking localRef, please use 

Pattern:
JniLocalReferenceManager manager(env); ==> manage the lifetime for all local refs in current function.
....
jstring yy = env->NewStringUTF(xx)

example code is here, https://github.com/project-chip/connectedhomeip/blob/master/src/controller/java/AndroidCallbacks.cpp#L93C5-L93C43

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:219]: For all error cases in these functions, should we call sendOTAQueryFailure across? In other places, it seems we siliently return or just show the error log?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:340]: both jAction and jDelayedActionTime are leaking local ref

File: src/controller/java/src/chip/devicecontroller/OTAProviderDelegate.java
Comment (yunhanw-google) [inline src/controller/java/src/chip/devicecontroller/OTAProviderDelegate.java:33]: please add doxygen for all java API, later we would generate these doxygen in github webpage..

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt:44]: This value must be the VendorID of the controller company. This was implemented like this because I thought that the same value as the value set in the code below should be set. Wouldn’t there be a separate UI?
https://github.com/project-chip/connectedhomeip/blob/f0e6fc99535f63d6e71e6c83e931d86d1b267987/examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt#L51

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp:587]: I see.

File: src/controller/java/AndroidDeviceControllerWrapper.cpp
Comment (joonhaengHeo) [reply src/controller/java/AndroidDeviceControllerWrapper.cpp:530]: I will add.

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp:561]: I modify.

File: src/controller/java/OTAProviderDelegateBridge.h
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.h:64]: I modify.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:182]: I added.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:340]: Add "chip::JniLocalReferenceManager manager(env);" in this function

File: src/controller/java/BdxOTASender.cpp
Comment (joonhaengHeo) [reply src/controller/java/BdxOTASender.cpp:195]: I add "chip::JniLocalReferenceManager manager(env)" code in this function

File: src/controller/java/BdxOTASender.cpp
Comment (joonhaengHeo) [reply src/controller/java/BdxOTASender.cpp:326]: I add chip::JniLocalReferenceManager manager(env)  in this function.

File: src/controller/java/BdxOTASender.cpp
Comment (joonhaengHeo) [reply src/controller/java/BdxOTASender.cpp:328]: This class can manage reference lifetime. Is there something I don't know?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:219]: I called sendOTAQueryFailure only when the app called an intentional Error. (e.g., when there is no transmitted data, BDX Protocol is not supported).
Is it correct to call it in other cases as well?

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt:118]: modified

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt:280]: I modifed

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt
Comment (yunhanw-google) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt:44]: yep, maybe separate UI would be cool, but the default value would be 0xFFF4. Assume developer wanna firstly use android chip-tool to exercise their company's product, they may wanna setup the UI as his company's vendor Id, and try to simulate as much as they can.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [reply src/controller/java/OTAProviderDelegateBridge.cpp:215]: The other side would timeout if there is no response here, which means client stucks somehow...
Also it would be good to report the error to controller application as well, instead of silently skipping it.

I would prefer generating right error to notify if we can.

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (yunhanw-google) [inline src/controller/java/CHIPDeviceController-JNI.cpp:575]: ThrowError?

File: src/controller/java/src/chip/devicecontroller/OTAProviderDelegate.java
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/OTAProviderDelegate.java:33]: I added for doxygen.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:247]: modified. All cases will be called callback function,.

File: src/controller/java/CHIPDeviceController-JNI.cpp
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp]: Added.

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/ChipClient.kt:44]: I added some UI (EditText) for changable vendorID.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:327]: when error happens, should we commandObj->AddResponse with bad status?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:350]: when error happens, should we commandObj->AddResponse with bad status?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: Understand. I modified.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:394]: command response with failure status?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:365]: command response with failure status?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:361]: command response with failure status?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: Modified.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:412]: This is simply a Notify and never responds.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: modified.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: modified.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [inline src/controller/java/OTAProviderDelegateBridge.cpp:153]: you add status here, then in exit from Line 322, you add response again?

Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:153]: I know that this code is only the status changing and does not send a response.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp:155]: I know that this code is only the status changing and does not send a response.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (yunhanw-google) [reply src/controller/java/OTAProviderDelegateBridge.cpp:153]: either you add status, commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Failure));
or 
commandObj->AddResponse(cachedCommandPath, errorResponse);

File: examples/android/CHIPTool/app/src/main/AndroidManifest.xml
Comment (nicelyjust) [inline examples/android/CHIPTool/app/src/main/AndroidManifest.xml:21]: it is dangerous permission，should be replaced by SAF

Comment (yunhanw-google) [reply examples/android/CHIPTool/app/src/main/AndroidManifest.xml:21]: SAF looks better here, thanks

Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/AndroidManifest.xml:21]: I agree with this comment. However, this part does not seem to be an important point in OTA implementation. After merging, how about modifying it using SAF?

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: Overall, it has been modified so that either AddStatus or Addresponse is called.

File: src/controller/java/OTAProviderDelegateBridge.cpp
Comment (joonhaengHeo) [reply src/controller/java/OTAProviderDelegateBridge.cpp]: Overall, it has been modified so that either AddStatus or Addresponse is called.

File: examples/android/CHIPTool/app/src/main/AndroidManifest.xml
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/AndroidManifest.xml:21]: I tried using SAF, but I am having a problem with the file not being read.

Of course, this permission is a dangerous permission, but I think this permission exists only for testing. In the case of actual commercial apps, I think this permission is not necessary because the file is downloaded from DCL and only the storage within the app is used. Is there really a need to delete this?

Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/AndroidManifest.xml]: I found that how to use SAF. I modified code and removed dangerous permission.

**Review by yunhanw-google — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**
